device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,220 @@
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.exceptions.DeviceOperationTimeoutException;
+import com.microsoft.azure.sdk.iot.device.exceptions.TransportException;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+
+/**
+ * This is the base class for DeviceClientManager and MultiplexingClientManager
+ * The shared logic for reconnection and handling status change callback exists here.
+ */
+@Slf4j
+public abstract class ClientManagerBase implements IotHubConnectionStatusChangeCallback, ConnectionStatusTracker
+{
+    protected static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 2;
+
+    // Initialize the connection status as DISCONNECTED
+    protected ConnectionStatus lastKnownConnectionStatus = ConnectionStatus.DISCONNECTED;
+
+    // Tracks connection status of the dependency connection (multiplexing client connection status for example in this case for DeviceClientManager)
+    protected ConnectionStatusTracker dependencyConnectionStatusTracker;
+
+    public abstract void openClient() throws IOException;
+    public abstract void closeClient() throws IOException;
+    public abstract String getClientId();
+
+    // Since the client manager is in charge of handling the connection status callback, this method is a no-op
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
```

Comment:
In .NET we'd make this abstract to ensure the inheriting class implemented it. Does Java have an analog?

Reply:
There is the same concept in java, I didn't make it abstract because there is nothing to override, we don't want the subclasses to do anything with this, we just want to make sure the call is delegated to the ClientManager and it doesn't do anything. we can discuss it in more depth if you like.